### PR TITLE
 Use set_type to implement type conversions in C++ API 

### PR DESCRIPTION
**Pull request by goldsborough**
This PR implements in-place type conversions in the C++ API/autogradpp. More precisely:
1. Implements `to(backend)`, `to(scalar_type)` and `to(type)` using `at::detail::set_type`, which changes a `Variable`'s `data()`'s type, as well as the `Variable`'s type itself. This effectively means type conversions can be performed in-place on Variables, where autogradpp had to design its API around re-initializing variables either in CUDA or CPU "mode".
2. Removes the `cuda_` flag from `Module`. The type of variables is now no longer a property of the module, but simply a property of the parameters of a module. This is equivalent to Python.
3. Removes `DefaultTensor` by either constructing new variables on the CPU, or constructing variables using dispatch from other variable's `.type()` member, to make them the same type.
4. Small naming change from `params_` to `parameters_`. I've kept the map a public member for now, but this will change in another PR that better encapsulates them.
5. Also adds `zero_grad()`, to be consistent with the Python API.

@ebetica @jgehring @apaszke @ezyang

### CODE REVIEW

File: torch/csrc/api/src/nn/modules/batchnorm.cpp [hunk 1]
```diff
@@ -3,15 +3,13 @@
...
-    bias = this->add(
-        Var(DefaultTensor(at::kFloat).tensor(num_features_), true), "bias");
+    weight = this->add(Var(at::CPU(at::kFloat).empty(num_features_)), "weight");
+    bias = this->add(Var(at::CPU(at::kFloat).empty(num_features_)), "bias");
   }
 
   if (stateful_) {
-    running_mean = Var(DefaultTensor(at::kFloat).zeros({num_features_}), false);
-    running_var = Var(DefaultTensor(at::kFloat).ones({num_features_}), false);
+    running_mean = Var(at::CPU(at::kFloat).zeros({num_features_}), false);
```
Comment (ezyang) [inline torch/csrc/api/src/nn/modules/batchnorm.cpp]: Why are these hard coded CPU now?

**Review by ezyang — APPROVED** [review]
This has BC implications for StarCraft and fbcode, iiuc

File: torch/csrc/api/src/nn/modules/batchnorm.cpp (hunk 1 above)
Comment (goldsborough) [reply torch/csrc/api/src/nn/modules/batchnorm.cpp]: Well I think it's usual (e.g. in Python PyTorch) that all tensors always start on the CPU, and one then calls `.cuda()` to move them over?

**Comment by ebetica** [comment]
Just a comment on 4. - I think we should keep things as public as possible to keep the API hackable. It was a conscious choice to keep everything public so it's easy to access anything users want to, but only maybe exposing the functions we want them to use in the docs / comments in the headers.

**Review by ebetica — APPROVED** [review]
